### PR TITLE
Deprecate `DistributedType` in favor of `StrategyType`

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Part of #9932 

Kudos to @ethanwharris for sharing the Enum deprecation trick! 🥳 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/enums.py [hunk 1]
```diff
@@ -73,30 +73,30 @@ def supported_types() -> List[str]:
         return [x.value for x in PrecisionType]
 
 
-class DistributedType(LightningEnum):
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/enums.py:76]: Same comment as on #10503 : https://github.com/PyTorchLightning/pytorch-lightning/pull/10503#discussion_r748284860

Comment (kaushikb11) [reply pytorch_lightning/utilities/enums.py:76]: Updated!

**Review by ethanwharris — APPROVED** [review]
Awesome, LGTM :smiley:

File: pytorch_lightning/utilities/enums.py [hunk 2]
```diff
@@ -73,17 +101,33 @@ def supported_types() -> List[str]:
...
     >>> # you can match the type with string
     >>> DistributedType.DDP == 'ddp'
     True
     >>> # which is case invariant
     >>> DistributedType.DDP2 in ('ddp2', )
     True
+
+    Deprecated since v1.6.0 and will be removed in v1.8.0.
+
+    Use `DistributedType` instead.
```
Comment (rohitgr7) [inline pytorch_lightning/utilities/enums.py:116]:
```suggestion
    Use `StrategyType` instead.
```

File: pytorch_lightning/utilities/enums.py [hunk 3]
```diff
@@ -37,6 +39,32 @@ def __hash__(self) -> int:
         return hash(self.value.lower())
 
 
+class OnAccessEnumMeta(EnumMeta):
```
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:42]: Make it protected

```suggestion
class _OnAccessEnumMeta(EnumMeta):
```

File: pytorch_lightning/utilities/enums.py [hunk 4]
```diff
@@ -37,6 +39,32 @@ def __hash__(self) -> int:
...
+    """Enum with a hook to run a function whenever a member is accessed.
+
+    Adapted from:
+    https://www.buzzphp.com/posts/how-do-i-detect-and-invoke-a-function-when-a-python-enum-member-is-accessed
+    """
+
+    def __getattribute__(cls, name):
```
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:49]: Add typing to the class

File: pytorch_lightning/utilities/enums.py [hunk 5]
```diff
@@ -188,3 +231,41 @@ def get_max_depth(mode: str) -> int:
     @staticmethod
     def supported_types() -> List[str]:
         return [x.value for x in ModelSummaryMode]
+
+
+class StrategyType(LightningEnum):
```
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:236]: Can we take advantage of this depreaction and make this protected?

```suggestion
class _StrategyType(LightningEnum):
```

`DistributedType` was never meant to be public. Users should always compare with the string value.

**Review by rohitgr7 — APPROVED** [review]
there are still some instances of it left inside notebook..

File: pytorch_lightning/utilities/enums.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/enums.py:236]: we have it publically available inside dqn example

Comment (kaushikb11) [reply pytorch_lightning/utilities/enums.py:236]: yeah, there could be cases of users using it

Comment (carmocca) [reply pytorch_lightning/utilities/enums.py:236]: > we have it publically available inside dqn example

That's not a good argument as it already accesses protected variables

`if self.trainer._distrib_type in {DistributedType.DP, DistributedType.DDP2}:`

users having to import and use enums is a source of boilerplate, it also means we are less limited in the changes we can do.

Users should always compare and use their string arguments, in that case

`in {"dp", "ddp"}`, as that's why the `LightningEnum` was introduced.

This is the case for the `stage` argument of `setup`

> there could be cases of users using it

This class is new, nobody is using it as you just wrote it, they will get the deprecation warning for the old one.

File: pytorch_lightning/utilities/enums.py (hunk 4 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/enums.py]: Updated

File: pytorch_lightning/utilities/enums.py (hunk 5 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/enums.py]: @carmocca Updated!

File: pytorch_lightning/utilities/enums.py [hunk 6]
```diff
@@ -98,17 +142,16 @@ def is_interactive_compatible(self) -> bool:
...
-    DDP_FULLY_SHARDED = "ddp_fully_sharded"
+    def __new__(cls, value):
+        member = str.__new__(cls, value)
+        member._on_access = member.deprecate
+        return member
+
+    def deprecate(self):
+        rank_zero_deprecation(
+            "`DistributedType` Enum has been deprecated in v1.6 and will be removed in v1.8."
+            " Use `_StrategyType` instead."
```
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:153]:
```suggestion
            " Use the string value `{self.value!r}` instead."
```

File: pytorch_lightning/utilities/enums.py [hunk 7]
```diff
@@ -98,17 +142,16 @@ def is_interactive_compatible(self) -> bool:
...
-    DDP2 = "ddp2"
-    DDP_CPU = "ddp_cpu"
-    DDP_SPAWN = "ddp_spawn"
-    TPU_SPAWN = "tpu_spawn"
-    DEEPSPEED = "deepspeed"
-    HOROVOD = "horovod"
-    DDP_SHARDED = "ddp_sharded"
-    DDP_SHARDED_SPAWN = "ddp_sharded_spawn"
-    DDP_FULLY_SHARDED = "ddp_fully_sharded"
+    def __new__(cls, value):
```
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:145]: typing

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/enums.py (hunk 7 above)
Comment (kaushikb11) [inline pytorch_lightning/utilities/enums.py:145]:
```suggestion
    def __new__(cls, value: str):
```

File: pytorch_lightning/utilities/enums.py [hunk 8]
```diff
@@ -98,17 +142,16 @@ def is_interactive_compatible(self) -> bool:
...
-    HOROVOD = "horovod"
-    DDP_SHARDED = "ddp_sharded"
-    DDP_SHARDED_SPAWN = "ddp_sharded_spawn"
-    DDP_FULLY_SHARDED = "ddp_fully_sharded"
+    def __new__(cls, value: str):
+        member = str.__new__(cls, value)
+        member._on_access = member.deprecate
+        return member
+
+    def deprecate(self):
```
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:150]:
```suggestion
    def __new__(cls, value: str) -> 'DistributedType':
        member = str.__new__(cls, value)
        member._on_access = member.deprecate
        return member

    def deprecate(self) -> None:
```

**Comment by four4fish** [comment]
Great! Thank you for fixing this. 
One big step towards https://github.com/PyTorchLightning/pytorch-lightning/issues/10422

File: pytorch_lightning/utilities/enums.py [hunk 9]
```diff
@@ -98,17 +142,16 @@ def is_interactive_compatible(self) -> bool:
...
-    DDP_SPAWN = "ddp_spawn"
-    TPU_SPAWN = "tpu_spawn"
-    DEEPSPEED = "deepspeed"
-    HOROVOD = "horovod"
-    DDP_SHARDED = "ddp_sharded"
-    DDP_SHARDED_SPAWN = "ddp_sharded_spawn"
-    DDP_FULLY_SHARDED = "ddp_fully_sharded"
+    def __new__(cls, value: str) -> "DistributedType":
+        member = str.__new__(cls, value)
+        member._on_access = member.deprecate
```
Comment (tchaton) [inline pytorch_lightning/utilities/enums.py:147]: Do we need this ?

File: pytorch_lightning/utilities/enums.py [hunk 10]
```diff
@@ -37,6 +39,32 @@ def __hash__(self) -> int:
...
+    def __getitem__(cls, name: str) -> Any:
+        member = super().__getitem__(name)
+        if member._on_access:
+            member._on_access()
+        return member
+
+    def __call__(cls, value: str, *args: Any, **kwargs: Any) -> Any:
+        obj = super().__call__(value, *args, **kwargs)
+        if isinstance(obj, Enum) and obj._on_access:
+            obj._on_access()
```
Comment (tchaton) [inline pytorch_lightning/utilities/enums.py:64]: Can we use `deprecate` instead ? Do you see other reasons why we would like a more general _on_access function ?

File: pytorch_lightning/utilities/enums.py [hunk 11]
```diff
@@ -188,3 +231,41 @@ def get_max_depth(mode: str) -> int:
...
+    def interactive_compatible_types() -> List["_StrategyType"]:
+        """Returns a list containing interactive compatible _StrategyTypes."""
+        return [
+            _StrategyType.DP,
+            _StrategyType.DDP_SPAWN,
+            _StrategyType.DDP_SHARDED_SPAWN,
+            _StrategyType.TPU_SPAWN,
+        ]
+
+    def is_interactive_compatible(self) -> bool:
```
Comment (tchaton) [inline pytorch_lightning/utilities/enums.py:269]: By `interactive`, do you mean notebooks ? Won't `is_notebook_compatible` more explicit ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/utilities/enums.py (hunk 11 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/enums.py:269]: Imo, it's an industry standard of it being called `interactive`

File: tests/trainer/test_trainer.py [hunk 12]
```diff
@@ -2161,40 +2161,40 @@ def training_step(self, batch, batch_idx):
...
         ),
         (
             dict(strategy=DDPSpawnPlugin(), gpus=2),
-            dict(_distrib_type=DistributedType.DDP_SPAWN, _device_type=DeviceType.GPU, num_gpus=2, num_processes=1),
+            dict(_distrib_type=_StrategyType.DDP_SPAWN, _device_type=DeviceType.GPU, num_gpus=2, num_processes=1),
         ),
         (
             dict(strategy=DDPPlugin(), num_processes=2, gpus=None),
-            dict(_distrib_type=DistributedType.DDP, _device_type=DeviceType.CPU, num_gpus=0, num_processes=2),
+            dict(_distrib_type=_StrategyType.DDP, _device_type=DeviceType.CPU, num_gpus=0, num_processes=2),
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:2180]: All these accesses in our tests will trigger deprecation warnings.

This goes against #9939 

Can you update all test usages to use the non-deprecated alternative (str) or catch the warnings with `pytest.deprecated_call(match=...)`?

**Review by carmocca — DISMISSED** [review]

File: pytorch_lightning/utilities/enums.py (hunk 9 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/enums.py]: Updated!

File: pytorch_lightning/utilities/enums.py (hunk 10 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/enums.py]: Updated!

**Pull request merged**